text/0244-eks-developer-preview-migration-path.md

```diff
@@ -0,0 +1,91 @@
+> If you are transitioning to this module from `aws-eks-experimental`, you might incur changes that require cluster replacements.
+
+# Implementation Plan
+
+1. Rename current module to `aws-eks-experimental` on the master branch.
+
+    - This will be released with a BREAKING CHANGE notice and cause users to change their dependency.
+
+2. Create a staging branch `eks-dev-preview` which contains `aws-eks` as a copy of `aws-eks-experimental`.
+
+    - This will be a long living branch we will branch out of to and merge to.
+
+3. Implement necessary changes onto the `eks-dev-preview` branch.
+
+    - We branch out of the `aws-dev-preview` branch to separate feature branches so that we can review the changes.
+
+4. Merge `eks-dev-preview` to master.
+
+    - This merge is the release of the developer preview module.
+
+# Rationale and Alternatives
+
+### Rename `eks.Cluster`
+
+Instead of spinning off to a new module, we can spin off to a new construct.
+
+That is, rename the current `eks.Cluster` to `eks.ExperimentalCluster` and implement everything we need in `eks.Cluster`. We have actually already done this once with `eks.LegacyCluster`.
+
+The benefit here is that users won't need to change dependencies. However, this change would also require a name change and duplication of any class that interacts with `eks.Cluster` since those classes may not work with both `eks.Cluster` and `eks.ExperimentalCluster`.
+
+This approach feels more fragile and doesn't provide much value since the user would still incur a breaking change. It would just be in the form of a class name change, instead of a module name change.
+
+### Rename new module
+
+Instead of renaming the current module, we can rename the new module to `aws-eks-dev-preview`. This would prevent the need for the detection of old module usages in the new module code.
+
+The problem here is that eventually we would need to rename `aws-eks-dev-preview` to `aws-eks`, which will be taken. So we would need to rename `aws-eks` as well, winding up in the same situation, except we now broke the old module again, after we promised not to.
+
+# Adoption Strategy
+
+When users upgrade their version of `aws-eks`, they will now receive the new intrusive version. This version will detect that the user had previously been using the old `aws-eks` (which is now `aws-eks-experimental`) module, and display the following error:
+
+```console
+The aws-eks module has undergone a name change, to use the previously named aws-eks module, please change your imports and dependencies to use aws-eks-experimental.
+If you'd like to use the new aws-eks module, add the following key to your cdk.json file: "@aws-cdk/aws-eks:dev-preview: true".
+```
+
+The user will then choose to either change imports to point to the `aws-eks-experimental` module, and continue working without interruption. Or, add the necessary feature flag to `cdk.json` and continue with the new module which will replace the cluster.
+
+# Future Possibilities
```

Comment:
One thing that's worth talking about is how do we make sure that the stuff we actually release as developer preview is working? It feels like there might be risk in accumulating many changes on the `eks-dev-preview` branch and then releasing all of them together as "developer preview" and then realizing something is horribly broken and having to release a breaking change to fix.

Reply:
Following our talk today, we decided to go with releasing an `@aws-cdk/aws-eks-next` that will incrementally introduce those changes. This will give us the option to use the module exactly as users would and validate it. We can also expose this module to a select few, or to those who choose to use it, knowing fully that it may introduce these types of changes frequently.

Reply:
Does that decision for `aws-eks-next` effectively invalidate the rest of this PR? Is this equivalent to the `Rename new module` alternative, except that we're not (yet) declaring the new module dev preview?

Reply:
Pretty much